libs/Matchmaking/src/Peer/PeerDiscoveryAgent.cs

```diff
@@ -571,13 +571,57 @@ class Client
         /// Protocol handler.
         Proto proto_;
 
+        Task updateTask_;
+        CancellationTokenSource updateCts_ = new CancellationTokenSource();
+        AutoResetEvent updateAvailable_ = new AutoResetEvent(false);
+
         internal Client(IPeerDiscoveryTransport net)
         {
             proto_ = new Proto(net);
             proto_.OnServerHello = OnServerAnnounce;
             proto_.OnServerByeBye = OnServerByeBye;
             proto_.OnServerReply = OnServerAnnounce;
             timer_ = new Timer(OnClientTimerExpired, null, Timeout.Infinite, Timeout.Infinite);
+
+            updateTask_ = Task.Run(() =>
+            {
+                const int MinIntervalMs = 200;
+                var timerExpired = new AutoResetEvent(true);
+                var handles = new WaitHandle[] { timerExpired, updateAvailable_ };
+                var tasksUpdated = new List<DiscoveryTask>();
+                var token = updateCts_.Token;
+                using (var updateTimer = new Timer(_ => { timerExpired.Set(); }))
+                {
+                    // Stop the wait if cancellation is requested.
+                    token.Register(() => updateAvailable_.Set());
+                    while (true)
+                    {
+                        WaitHandle.WaitAll(handles);
```

Comment:
Don't you need to pass the token here?

Reply:
No, this is Wait**All**. If the cancellation is issued I wake up this call by calling `updateAvailable_.Set()` (line 596), but I don't want to wait for cancellation every time 😄 

Will add comments to clarify this.

Reply:
Talked IRL - WaitAll(Task[], CancellationToken) looks like what we need

Reply:
Re-talked IRL - WaitHandle.WaitAll doesn't take a CancellationToken (for some reason).

We established that the minimum interval is not that useful and removed it.